test/performance/cicd/perf-test.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package perf
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	ksapi "github.com/kubestellar/kubestellar/api/control/v1alpha1"
+	"github.com/kubestellar/kubestellar/test/util"
+)
+
+const (
+	ns = "nginx"
+)
+
+var _ = ginkgo.Describe("performance tests", func() {
+	ginkgo.BeforeEach(func() {
+		util.CleanupWDS(ctx, wds, ksWds, ns)
+	})
+
+	ginkgo.Context("scalability", func() {
+		ginkgo.It("create a bindingpolicy and a single deployment", func() {
+			const numDeployments = 10
+			sumBinding, sumManifestwork, sumDeploymentWDS, sumDeploymentWEC, sumTotal := 0, 0, 0, 0, 0
+
+			for i := 0; i < numDeployments; i++ {
+				util.CreateBindingPolicy(ctx, ksWds, "performance",
+					[]metav1.LabelSelector{
+						{MatchLabels: map[string]string{"name": "cluster1"}},
+					},
+					[]ksapi.DownsyncObjectTest{
+						{ObjectSelectors: []metav1.LabelSelector{
+							{MatchLabels: map[string]string{"app.kubernetes.io/name": "nginx"}},
+						}}})
+				start := time.Now()
+				util.CreateDeployment(ctx, wds, ns, fmt.Sprint(i),
+					map[string]string{
+						"app.kubernetes.io/name": "nginx",
+					})
+				util.ValidateNumDeployments(ctx, wec1, ns, 1)
+
+				timeDeploymentWDS := int(util.GetDeploymentTime(ctx, wds, ns, fmt.Sprint(i)).Sub(start).Seconds())
+				timeBinding := int(util.GetBindingTime(ctx, ksWds, "performance").Sub(start).Seconds())
+				timeManifestwork := int(util.GetManifestworkTime(ctx, its, "cluster1", "performance-wds1").Sub(start).Seconds())
+				timeDeploymentWEC := int(util.GetDeploymentTime(ctx, wec1, ns, fmt.Sprint(i)).Sub(start).Seconds())
+				elapsed := int(time.Since(start).Seconds())
```

Comment:
Why not use `util.GetDeploymentTime(ctx, wec1, ns, fmt.Sprint(i))` as the endpoint rather than `time.Now()`?

Reply:
I wanted to have an overall timestamp that is collected by the test itself rather than obtained indirectly from the object timestamps.   Just in case there is some additional overhead that is not accounted by the object creation.

Reply:
> Just in case there is some additional overhead

So collect both timestamps and report whether there is additional overhead.

Reply:
I am not seeing a `util.GetDeploymentTime` in test/util/util.go?

Reply:
I see this PR adding GetDeploymentTime, and using it in the line above the one being discussed here.

Reply:
Sorry my bad.  It's been so long I forgot about this function and diffed against main :-|  I am actually reporting both as you are suggesting above.  But I don't print out a specific message saying there is an additional overhead.  If you think one is needed I can add one.

Reply:
> Just in case there is some additional overhead that is not accounted by the object creation.

There certainly _is_ additional activity between object creation and `elapsed := int(time.Since(start).Seconds())`. Look at those preceding statements that do apiserver queries, and I do not know whether `gomega.Eventually` delays before the first call to its given function. And there is the stuff that happens between the apiserver creating the object and the return from `util.ValidateNumDeployments`.

>  But I don't print out a specific message saying there is an additional overhead. If you think one is needed I can add one.

I do not think that will be necessary.

BTW, I was surprised to learn that the times reported in one line are cumulative from left to right; labeling them with things like "manifestwork=" suggests, to me, that the number shown is just for the delta from the previous time point.

Reply:
Are we good here or would you want any particular change?  Like change the cumulative into delta?

Reply:
Cumulative is good enough, as long as it is clear to the reader.